third_party/move/move-model/src/metadata.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::bail;
+use move_command_line_common::env::read_bool_env_var;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{
+    fmt,
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+const UNSTABLE_MARKER: &str = "-unstable";
+
+/// Metadata about a compilation result. To maintain serialization
+/// compatibility, this uses a free-form string to represent compiler version
+/// and language version.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct CompilationMetadata {
+    /// The version of the compiler, as a string.
+    pub compiler_version: String,
+    /// The version of the language, as a string.
+    pub language_version: String,
+    /// A flag indicating whether, at time of creation, the compilation
+    /// result was considered as unstable. Unstable code may have restrictions
+    /// for deployment on production networks.
+    pub unstable: bool,
+}
+
+impl CompilationMetadata {
+    pub fn new(compiler_version: CompilerVersion, language_version: LanguageVersion) -> Self {
+        Self {
+            compiler_version: compiler_version.to_string(),
+            language_version: language_version.to_string(),
+            unstable: compiler_version.unstable() || language_version.unstable(),
+        }
+    }
+
+    pub fn compiler_version(&self) -> anyhow::Result<CompilerVersion> {
+        CompilerVersion::from_str(&self.compiler_version)
+    }
+
+    pub fn language_version(&self) -> anyhow::Result<LanguageVersion> {
+        LanguageVersion::from_str(&self.language_version)
+    }
+
+    /// Returns true of the compilation was created as unstable.
+    pub fn created_as_unstable(&self) -> bool {
+        self.unstable
+    }
+}
+
+/// Represents a compiler version
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd, Ord)]
+pub enum CompilerVersion {
+    /// The legacy v1 Move compiler
+    V1,
+    /// The v2 compiler, starting with 2.0. Each new released version of the compiler
+    /// should get an enum entry here.
+    V2_0,
+}
+
+impl Default for CompilerVersion {
+    /// We allow the default to be set via an environment variable.
+    fn default() -> Self {
+        static MOVE_COMPILER_V2: Lazy<bool> = Lazy::new(|| read_bool_env_var("MOVE_COMPILER_V2"));
+        if *MOVE_COMPILER_V2 {
+            Self::V2_0
+        } else {
+            Self::V1
+        }
+    }
+}
+
+impl FromStr for CompilerVersion {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // Strip unstable marker as it is not relevant for parsing.
+        let s1 = s.replace(UNSTABLE_MARKER, "");
+        match s1.as_str() {
+            // For legacy reasons, also support v1 and v2
+            "1" | "v1" => Ok(Self::V1),
+            "2" | "v2" | "2.0" => Ok(Self::V2_0),
```

Comment:
- Its always only two digits. But yes, 2 is 2.latest. We can flesh this out as we move.
- The v1/v2 is really only for cli api compatibility. We would need to find where its used (some github actions perhaps) but this can be done in another PR.

Reply:
Ok, I clarified the versioning scheme for both compiler and language version.

Reply:
+1

Let's just use the standard 3-part Semantic versioning.